### ISSUE TITLE
Unexpected initial value with jit

### ISSUE THREAD

**Issue by Mskxn** [numba/numba#9657]
## Reporting a bug

- [x] I have tried using the latest released version of Numba (most recent is
  visible in the release notes
  (https://numba.readthedocs.io/en/stable/release-notes-overview.html).
- [x] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

I think this maybe a bug. 
Though cpython throws an exception when `flag` is False, according to https://github.com/numba/numba/issues/9647#issue-2400041502, x should be allocated before executing and init with zero, is also acceptable. 
But actually, the following code returns 2 when `flag` is False.
```python
from numba import jit

@jit
def f(flag):
    if flag:
        x = 2
    return x

r = f(False)
print(r) # expect Exception (or 0) but got 2
```

**Comment by gmarkall** [comment]
It's not exactly well-signposted in the docs, but variable liveness is not checked by Numba: https://numba.readthedocs.io/en/stable/reference/pysemantics.html#zero-initialization-of-variables

The docs could do with an edit / update to make this deviation a little more prominent / clear.

**Comment by esc** [comment]
Yeah, this looks like a bug. My understanding is that the above code should work like:

```
from numba import jit

@jit
def f(flag):
    x = 0
    if flag:
        x = 2
    return x

print(f(False))
print(f(True))
```

**Comment by stuartarchibald** [comment]
Think this is a duplicate of #7338, essentially, type inference propagates `2` as the value for `x` across the `phi` node without acknowledging the undefined nature of one of the incoming values.
